apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/MbiCache.java

```diff
@@ -159,37 +191,53 @@ static class DatabaseLookupFunction implements LookupFunction {
      */
     @VisibleForTesting
     @Override
-    public Mbi lookupMbi(String mbi) {
+    public Mbi lookupMbiImpl(String mbi) {
+      Mbi result = null;
+      int loopNumber = 0;
       int retryNumber = 0;
```

Comment:
I'm trying to understand the conditions where these two counters diverge. It would have to be a case where readOrInsertIfMissing() did not throw, but returned null. That implies that in readOrInsertIfMissing() we successfully committed but record was still null? I'm missing something here. If you can explain to me here and put in a comment about it that might also help someone else later on.

Reply:
First loop just does a query and returns a value if the database already has one.  Other loops try to insert a new record but that can fail if there is a conflict in the transactions.  Each iteration imposes a random wait to reduce the chances of a conflict happening.

Reply:
Loops necessitated by a conflict are retries.

Reply:
I agree the logic was a bit convoluted.  The ultimate goal was just to be able to track a miss (i.e. record had to be inserted because there was none in the database) vs a hit.  I have changed the logic to make it more obvious what is happening.

Reply:
thanks!